aptos-move/aptos-release-builder/src/simulate.rs

```diff
@@ -0,0 +1,500 @@
+struct SimulationStateView<'a, S> {
+    remote: &'a S,
+    states: Mutex<HashMap<StateKey, Option<StateValue>>>,
+}
+
+impl<'a, S> SimulationStateView<'a, S>
+where
+    S: StateView,
+{
+    fn set_state_value(&self, state_key: StateKey, state_val: StateValue) {
+        self.states
+            .lock()
+            .unwrap()
+            .insert(state_key, Some(state_val));
+    }
+
+    #[allow(dead_code)]
+    fn remove_state_value(&mut self, state_key: &StateKey) {
+        self.states.lock().unwrap().remove(state_key);
+    }
+
+    fn apply_write_set(&self, write_set: WriteSet) {
+        let mut states = self.states.lock().unwrap();
+
+        for (state_key, write_op) in write_set {
+            match write_op.as_state_value() {
+                None => {
+                    states.remove(&state_key);
+                },
+                Some(state_val) => {
+                    states.insert(state_key, Some(state_val));
+                },
+            }
+        }
+    }
+
+    #[allow(dead_code)]
+    fn read_resource<T: MoveResource>(&self, addr: &AccountAddress) -> T {
+        let data_blob = self
+            .get_state_value_bytes(
+                &StateKey::resource_typed::<T>(addr).expect("failed to create StateKey"),
+            )
+            .expect("account must exist in data store")
+            .unwrap_or_else(|| panic!("Can't fetch {} resource for {}", T::STRUCT_NAME, addr));
+
+        bcs::from_bytes(&data_blob).expect("failed to deserialize resource")
+    }
+}
+
+impl<'a, S> TStateView for SimulationStateView<'a, S>
+where
+    S: StateView,
+{
+    type Key = StateKey;
+
+    fn get_state_value(&self, state_key: &Self::Key) -> StateStoreResult<Option<StateValue>> {
+        if let Some(res) = self.states.lock().unwrap().get(state_key) {
+            return Ok(res.clone());
+        }
+        self.remote.get_state_value(state_key)
+    }
+
+    fn get_usage(&self) -> StateStoreResult<StateStorageUsage> {
+        panic!("not supported")
+    }
+
+    fn as_in_memory_state_view(&self) -> InMemoryStateView {
+        panic!("not supported")
+    }
+}
+
+/***************************************************************************************************
+ * Move Patches
+ *
+ **************************************************************************************************/
+static MODULE_ID_CREATE_SIGNER: Lazy<ModuleId> = Lazy::new(|| {
+    ModuleId::new(
+        AccountAddress::ONE,
+        Identifier::new("create_signer").unwrap(),
+    )
+});
+
+static MODULE_ID_APTOS_GOVERNANCE: Lazy<ModuleId> = Lazy::new(|| {
+    ModuleId::new(
+        AccountAddress::ONE,
+        Identifier::new("aptos_governance").unwrap(),
+    )
+});
+
+/// Makes `aptos_framework::create_signer::create_signer` a public function so that
+/// it can be used freely in the simulation.
+fn patch_create_signer(state_view: &SimulationStateView<impl StateView>) -> Result<()> {
```

Comment:
I like this patching and wondering if we can use these mocks beyond this tool to test private functions or hard to test things, like prologue or epilogue in AptosVM. There, we can "inject code in prologue/epilogue" to ensure it gets published and we can control it/assert.

Reply:
Like this idea. I've actually been thinking that if we come up with a way to allow these mock versions to be written in Move, they could have a lot more use.

CFG-cased conditional compilation would be sufficient
```
#[cfg(mock)]
fun do_something(..) { .. }

#[cfg(not(mock))]
fun do_something(..) { .. }
```
but it can be a bit wasteful as we have to recompile everything. (May not be a real problem though)

Even better I would love to have a way to write the mock version separately, and only compile that function against the original code.